src/libutil/ustring.cpp

```diff
@@ -89,11 +94,191 @@ typedef unordered_map_concurrent <string_view, ustring::TableRep *, Strutil::Str
+        memory_usage += (mask + 1) * sizeof(ustring::TableRep*);
+
+        ustring::TableRep** new_entries = static_cast<ustring::TableRep**>(calloc(new_mask + 1, sizeof(ustring::TableRep*)));
+        size_t to_copy = num_entries;
+        for (size_t i = 0; to_copy != 0; i++) {
+            if (entries[i] == 0)  continue;
+            size_t pos = entries[i]->hashed & new_mask, dist = 0;
+            for (;;) {
+                if (new_entries[pos] == 0)
+                    break;
+                ++dist;
+                pos = (pos + dist) & new_mask; // quadratic probing
+            }
+            new_entries[pos] = entries[i];
+            to_copy--;
+        }
+
+        free(entries);
+        entries = new_entries;
+        mask    = new_mask;
+    }
+
+    ustring::TableRep* make_rep(string_view str, size_t hash) {
+        char* repmem = pool_alloc(sizeof(ustring::TableRep) + str.length() + 1);
+        return new (repmem) ustring::TableRep (str, hash);;
+    }
+
+    char* pool_alloc(size_t len) {
+        if (len >= POOL_SIZE) {
+            memory_usage += len;
+            return (char*) malloc(len); // no need to try and use the pool
+        }
+        if (pool_offset + len > POOL_SIZE) {
+            // NOTE: old pool will leak - this is ok because ustrings cannot be freed
+            memory_usage += POOL_SIZE;
+            pool = (char*) malloc(POOL_SIZE);
+            pool_offset = 0;
+        }
+        char* result = pool + pool_offset;
+        pool_offset += len;
+        return result;
+    }
+
+    OIIO_CACHE_ALIGN ustring_mutex_t mutex;
+    size_t mask;
+    ustring::TableRep** entries;
+    size_t num_entries;
+    char* pool;
+    size_t pool_offset;
+    size_t memory_usage;
+    OIIO_CACHE_ALIGN size_t num_lookups;
+};
+
+#if USE_CUSTOM_MAP == 1
+typedef TableRepMap<1 << 20, 4 << 20> UstringTable;
+#elif USE_CUSTOM_MAP == 2
+struct UstringTable {
+    const char* lookup(string_view str, size_t hash) {
+        return whichbin(hash).lookup(str, hash);
+    }
+
+    const char* insert(string_view str, size_t hash) {
+        return whichbin(hash).insert(str, hash);
+    }
+
+    size_t get_memory_usage() {
+        size_t mem = 0;
+        for (int i = 0; i < NUM_BINS; i++)
+            mem += bins[i].get_memory_usage();
+        return mem;
+    }
+
+    size_t get_num_entries() {
+        size_t num = 0;
+        for (int i = 0; i < NUM_BINS; i++)
+            num += bins[i].get_num_entries();
+        return num;
+    }
+
+    size_t get_num_lookups() {
+        size_t num = 0;
+        for (int i = 0; i < NUM_BINS; i++)
+            num += bins[i].get_num_lookups();
+        return num;
+    }
+
+private:
+    enum { NUM_BINS = 32 }; // NOTE: must be power of 2
+
+    typedef TableRepMap<(1 << 20) / NUM_BINS, (4 << 20) / NUM_BINS> Bin;
+
+    Bin bins[NUM_BINS];
+
+    Bin& whichbin(size_t hash) {
+        size_t h = (size_t) murmur::fmix (uint64_t(hash));  // scramble again
```

Comment:
Since you're masking the 64 bit hash to use low bits only (based on table size), why should this extra call to fmix() be superior to the less-expensive technique of using higher bits from the hash to determine the bin? If it's a strong hash, those should be uncorrelated to the low bits you use for the slot within the bin, no?